spyder/widgets/variableexplorer/dataframeeditor.py

```diff
@@ -611,28 +563,226 @@ def copy(self):
+            else:
+                return Qt.AlignRight | Qt.AlignVCenter
+        if role != Qt.DisplayRole:
+            return None
+        if self.axis == 1 and self._shape[1] <= 1:
+            return None
+        orient_axis = 0 if orientation == Qt.Horizontal else 1
+        if self.model.header_shape[orient_axis] > 1:
+            header = section
+        else:
+            header = self.model.header(self.axis, section)
+            if isinstance(header, TEXT_TYPES):
+                # Get the proper encoding of the text in the header.
+                # Fixes Issue 3896
+                if not PY2:
+                    try:
+                        header = header.encode('utf-8')
+                        coding = 'utf-8-sig'
+                    except UnicodeEncodeError:
+                        coding = encoding.get_coding(header)
+                else:
+                    coding = encoding.get_coding(header)
+                return to_text_string(header, encoding=coding)
+            header = to_text_string(header)
+        return header
+
+    def data(self, index, role):
+        if not index.isValid() or \
+           index.row() >= self._shape[0] or \
+           index.column() >= self._shape[1]:
+            return None
+        row, col = ((index.row(), index.column()) if self.axis == 0
+                    else (index.column(), index.row()))
+        if role == Qt.BackgroundRole:
+            prev = self.model.header(self.axis, col - 1, row) if col else None
+            cur = self.model.header(self.axis, col, row)
+            return self._palette.midlight() if prev != cur else None
+        if role != Qt.DisplayRole:
+            return None
+        if self.axis == 0 and self._shape[0] <= 1:
+            return None
+        return to_text_string(self.model.header(self.axis, col, row))
+
+
+class DataFrameLevelModel(QAbstractTableModel):
+    """
+    Data Frame level class.
+
+    This class model the levels in the DataFrameEditor. When using multiindex
+    this model creates the labels for the index/header as Index i for each
+    section in the index/header
+
+    Taken from the gtabview project(Level4ExtModel).
+    For more information please see:
+    https://github.com/wavexx/gtabview/blob/master/gtabview/viewer.py
+    """
+
+    def __init__(self, model, palette, font):
+        super(DataFrameLevelModel, self).__init__()
+        self.model = model
+        self._background = palette.dark().color()
+        if self._background.lightness() > 127:
+            self._foreground = palette.text()
+        else:
+            self._foreground = palette.highlightedText()
+        self._palette = palette
+        font.setBold(True)
+        self._font = font
+
+    def rowCount(self, index=None):
+        return max(1, self.model.header_shape[0])
+
+    def columnCount(self, index=None):
+        return max(1, self.model.header_shape[1])
+
+    def headerData(self, section, orientation, role):

```

Comment:
Missing docstring

Reply:
Since the `headerData` is re-written, maybe changes of #3812 (tooltip display) could be taken into account as part of this PR?